--- ../slang-1.2.2/src/_slang.h	Fri Apr 24 08:17:04 1998
+++ src/_slang.h	Mon Mar 27 23:52:45 2000
@@ -14,6 +14,22 @@
 #include "jdmacros.h"
 #include "sllimits.h"
 
+/* LERMEN, 2000327 for dosemu-1.0.1
+ *
+ * Kludge to make libslang.a includeable in both, libc5 and glibc > 2.1
+ * The problem is _IO_std* no longer exported in glibc-2.1.
+ * In libc5 std{in,out,err} are macros defining an absolute pointer
+ * within libc, in glibc-2.1 they remain variables of type (FILE *)
+ * We just undef std{in,out,err}, /usr/include/stdio.h of libc5 has them
+ * also declared as
+ *   extern FILE *stdin, *stdout, *stderr;
+ * (mainly for debuger purposes, well but they are there;-)
+ */
+#undef stdin
+#undef stdout
+#undef stderr
+
+
 #ifdef VMS
 # define SLANG_SYSTEM_NAME "_VMS"
 #else
